marimo/_plugins/ui/_impl/chat/types.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import annotations
+
+import abc
+from dataclasses import dataclass
+from typing import List, Literal, Optional, TypedDict
+
+
+class ChatAttachmentDict(TypedDict):
+    name: str
+    content_type: str
+    url: str
+
+
+class ChatMessageDict(TypedDict):
+    role: Literal["user", "assistant", "system"]
+    content: str
+    attachments: Optional[List[ChatAttachmentDict]]
+
+
+@dataclass
+class ChatAttachment:
+    """
+    @public
+    """
+
+    # The name of the attachment, usually the file name.
+    name: str
+
+    # A string indicating the [media type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type).
+    # By default, it's extracted from the pathname's extension.
+    content_type: str
+
+    # The URL of the attachment. It can either be a URL to a hosted file or a
+    # [Data URL](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs).
+    url: str
+
+
+@dataclass
+class ChatMessage:
+    """
+    @public
+    """
+
+    role: Literal["user", "assistant", "system"]
+    content: str
+    attachments: Optional[List[ChatAttachment]] = None
+
+
+@dataclass
+class ChatModelConfig:
+    """
+    @public
+    """
+
+    max_tokens: Optional[int] = None
+    temperature: Optional[float] = None
+    top_p: Optional[float] = None
+    top_k: Optional[int] = None
+    frequency_penalty: Optional[float] = None
+    presence_penalty: Optional[float] = None
+
+
+# class ModelRegistry(abc.ABC):
+#     @abc.abstractmethod
+#     def get_model(self, model_name: str):
+#         pass
+
+#     @abc.abstractmethod
+#     def has_model(self):
+#         pass
+
+
+class ChatModel(abc.ABC):
+    """
+    @public
```

Comment:
What is this?

Reply:
this is a ts feature, my intention is to let us know that this is public api and must be backward compat